### PR TITLE
Dynamic goal generation for navigation

### PR DESCRIPTION
**Pull request by ojasjoshi**
*Description of changes:
This PR adds a random valid goal generation feature to enable navigation in newer maps and worlds. It replaces the current approach of feeding fixed user-set goals to the route manager via a YAML file with dynamic valid random goals fed from the goal generator. 

Note: Validity of a goal implies that the navigation stack can find a valid global path between the current move_base pose and the goal pose.

*Tests:
The build and run tests were run locally on docker containers with ROS Kinetic + Gazebo7 and ROS Melodic + Gazebo9 configurations. The accuracy of the goal generation was tested on new 0.05, 0.1 resolution maps for bookstore,small_house and small_warehouse worlds as well as on the currently available maps available in this repository for the respective worlds.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: simulation_ws/src/aws_robomaker_simulation_common/nodes/route_manager [hunk 1]
```diff
@@ -48,55 +200,58 @@ class RouteManager():
...
                 self.client.send_goal(current_goal)
-
-                if not self.client.wait_for_result():
-                    rospy.logerr("Move server not ready, will try again...")
-                else:
-                    if self.client.get_result():
-                        rospy.loginfo("Goal done: %s", current_goal)
+                
+                try:
+                    #wait 5sec for global plan to be published. If not, scan for a new goal..
```
Comment (samuelgundry) [inline simulation_ws/src/aws_robomaker_simulation_common/nodes/route_manager:282]: Why is this necessary? Was something breaking?

File: simulation_ws/src/aws_robomaker_simulation_common/nodes/route_manager [hunk 2]
```diff
@@ -48,55 +200,58 @@ class RouteManager():
...
+                except rospy.exceptions.ROSException:
+                    self.bad_goal_counter += 1
+                    rospy.logwarn("No plan found for goal. Scanning for a new goal...")
+                
                 rate.sleep()
-            except StopIteration:
-                rospy.loginfo("No goals, stopping route manager")
+            
+            except ValueError:
+                rospy.loginfo("No valid goal was found in the map, stopping route manager")
```
Comment (samuelgundry) [inline simulation_ws/src/aws_robomaker_simulation_common/nodes/route_manager]: Add the error message to the loginfo to make it easier to debug.  For instance, `"No valid goal was found in the map, stopping route manager. Error was: %s"

File: simulation_ws/src/cloudwatch_simulation/launch/bookstore_turtlebot_navigation.launch [hunk 3]
```diff
@@ -32,8 +32,6 @@
 
   <!-- Send navigation route goals to the robot in random order -->
   <group if="$(arg follow_route)">
-    <node pkg="aws_robomaker_simulation_common" type="route_manager" name="route_manager" output="screen">
-      <rosparam file="$(find aws_robomaker_bookstore_world)/routes/route.yaml" command="load"/> 
-    </node>
+    <node pkg="aws_robomaker_simulation_common" type="route_manager" name="route_manager" output="screen"/>
```
Comment (samuelgundry) [inline simulation_ws/src/cloudwatch_simulation/launch/bookstore_turtlebot_navigation.launch]: Let's not change existing interfaces , it's less disruptive and less surface area for testing. 

 I suggest keeping these as-is, so the route_manager can load a file to send new points or use the /map topic. So there's a "point provider".   Then use the rosparam config to toggle the mode. There is already a `mode` in the .yaml file, so just use this. And have a new separate launch file for using loading the random map config yaml.

**Review by samuelgundry — CHANGES_REQUESTED** [review]
This is great.  However, to keep the change less impactful, at least initially, keep the existing behaviour to make testing easier and less disruptive. And add the new capabilities. One reason is downstream users of the existing launch files may expect certain behaviour.

File: simulation_ws/src/aws_robomaker_simulation_common/nodes/route_manager (hunk 1 above)
Comment (ojasjoshi) [reply simulation_ws/src/aws_robomaker_simulation_common/nodes/route_manager:282]: this is the best sanity check to see if the navigation stack found a global plan to the goal, hence making the goal "valid"

File: simulation_ws/src/aws_robomaker_simulation_common/nodes/route_manager (hunk 2 above)
Comment (ojasjoshi) [reply simulation_ws/src/aws_robomaker_simulation_common/nodes/route_manager]: Done

File: simulation_ws/src/cloudwatch_simulation/launch/bookstore_turtlebot_navigation.launch (hunk 3 above)
Comment (ojasjoshi) [reply simulation_ws/src/cloudwatch_simulation/launch/bookstore_turtlebot_navigation.launch]: Restored the launch files to keep the existing interfaces same.. the dynamic route manager will be added to a new launch file (which will be part of Ansons PR)

**Review by samuelgundry — COMMENTED** [review]
I'm good with the changes. Did you want to add a *new* launch file to invoke the `dynamic` route manager?

**Comment by ojasjoshi** [comment]
Adding a generic launch file for dynamic goals isn't very user friendly. Mainly because,
1. for worlds outside aws-robotics(custom worlds), it would require that the navigation launch file and occupancy maps exist. (in which case I don't see cloudwatch SA being used only for purposes of getting dynamic goals.)
2. for aws-robotics worlds, there's not much value addition from a generic launch file if we can use the existing launch files while keeping the original usage and interfaces undisturbed.

The current version,
1. provides a argument group which would run dynamic goals navigation via the same launch files
2. keeps the default usage and any downstream interfaces same

**Review by mm318 — APPROVED** [review]

File: simulation_ws/src/aws_robomaker_simulation_common/nodes/route_manager [hunk 4]
```diff
@@ -65,38 +240,57 @@ class RouteManager():
...
         goal.target_pose.pose.orientation = Quaternion(**pose['pose']['orientation'])
         return goal
 
+
     def route_forever(self):
         rate = rospy.Rate(1)
         while not rospy.is_shutdown():
             try:
+                if self.bad_goal_counter > 10:
+                    raise ValueError("Bad goal limit exceeded")
```
Comment (mm318) [inline simulation_ws/src/aws_robomaker_simulation_common/nodes/route_manager]: Nitpick: I don't think this should be a `raise` or a `try-except` block, I think it should just be an `if-else` block. You can move the `try` block below up to include `to_move_goal()` and add another `except` branch to it for `ValueError`.

Comment (ojasjoshi) [reply simulation_ws/src/aws_robomaker_simulation_common/nodes/route_manager]: agreed, throwing error is not ideal behavior for this

File: simulation_ws/src/aws_robomaker_simulation_common/nodes/route_manager [hunk 5]
```diff
@@ -65,39 +240,58 @@ class RouteManager():
...
-                    if self.client.get_result():
-                        rospy.loginfo("Goal done: %s", current_goal)
-                rate.sleep()
-            except StopIteration:
-                rospy.loginfo("No goals, stopping route manager")
+            if self.bad_goal_counter > 10:
+                rospy.loginfo("Stopping route manager due to too many bad goals. Check that your occupancy map has 'Trinary' value representation and is not visually noisy/incorrect")
                 return
+            else:
+                try:
```
Comment (mm318) [inline simulation_ws/src/aws_robomaker_simulation_common/nodes/route_manager:272]: I think having nested `try-except` blocks is confusing. Either combine them into one, or have one only encapsulate the call to `to_move_goal()` and the other only encapsulate `wait_for_message()`.

**Review by mm318 — APPROVED** [review]

**Review by samuelgundry — APPROVED** [review]
